pkg/remotewrite/prometheus.go

```diff
@@ -23,121 +29,109 @@ func MapTagSet(t *metrics.TagSet) []*prompb.Label {
-				Name:  "__name__",
-				Value: fmt.Sprintf("%s%s_max", defaultMetricPrefix, series.Metric.Name),
-			}),
-			Samples: []*prompb.Sample{
-				{
-					Value:     sink.Max,
-					Timestamp: timestamp,
-				},
-			},
-		},
-		{
-			Labels: append(labels, &prompb.Label{
-				Name:  "__name__",
-				Value: fmt.Sprintf("%s%s_avg", defaultMetricPrefix, series.Metric.Name),
-			}),
-			Samples: []*prompb.Sample{
-				{
-					Value:     sink.Avg,
-					Timestamp: timestamp,
-				},
-			},
-		},
-		{
-			Labels: append(labels, &prompb.Label{
-				Name:  "__name__",
-				Value: fmt.Sprintf("%s%s_med", defaultMetricPrefix, series.Metric.Name),
-			}),
-			Samples: []*prompb.Sample{
-				{
-					Value:     sink.Med,
-					Timestamp: timestamp,
-				},
-			},
-		},
-		{
-			Labels: append(labels, &prompb.Label{
-				Name:  "__name__",
-				Value: fmt.Sprintf("%s%s_p90", defaultMetricPrefix, series.Metric.Name),
-			}),
-			Samples: []*prompb.Sample{
-				{
-					Value:     sink.P(0.9),
-					Timestamp: timestamp,
-				},
-			},
-		},
-		{
-			Labels: append(labels, &prompb.Label{
-				Name:  "__name__",
-				Value: fmt.Sprintf("%s%s_p95", defaultMetricPrefix, series.Metric.Name),
-			}),
-			Samples: []*prompb.Sample{
-				{
-					Value:     sink.P(0.95),
-					Timestamp: timestamp,
-				},
-			},
-		},
+	tg := &trendAsGauges{
+		series:    make([]*prompb.TimeSeries, 0, 8),
+		labels:    MapSeries(series),
+		timestamp: t.UnixMilli(),
+	}
+	tg.FindNameIndex()
+
+	tg.Append("count", float64(sink.Count))
+	tg.Append("sum", sink.Sum)
+	tg.Append("min", sink.Min)
+	tg.Append("max", sink.Max)
+	tg.Append("avg", sink.Avg)
+	tg.Append("med", sink.P(0.5))
+	tg.Append("p90", sink.P(0.9))
+	tg.Append("p95", sink.P(0.95))
+	return tg.series
+}
+
+type trendAsGauges struct {
+	// series is the slice of the converted TimeSeries.
+	series []*prompb.TimeSeries
+
+	// labels are the shared labels between all the Gauges.
+	labels []*prompb.Label
+
+	// timestamp is the shared timestamp in ms between all the Gauges.
+	timestamp int64
+
+	// ixname is the slice's index
+	// of the __name__ Label item.
+	//
+	// 16 bytes should be enough for the max length
+	// an higher value will probably generate
+	// serious issues in other places.
+	ixname uint16
+}
+
+func (tg *trendAsGauges) Append(suffix string, v float64) {
+	ts := &prompb.TimeSeries{
+		Labels:  make([]*prompb.Label, len(tg.labels)),
+		Samples: make([]*prompb.Sample, 1),
+	}
+	for i := 0; i < len(tg.labels); i++ {
+		ts.Labels[i] = &prompb.Label{
+			Name:  tg.labels[i].Name,
+			Value: tg.labels[i].Value,
+		}
+	}
+	ts.Labels[tg.ixname].Value += "_" + suffix
+
+	ts.Samples[0] = &prompb.Sample{
+		Timestamp: tg.timestamp,
+		Value:     v,
+	}
+	tg.series = append(tg.series, ts)
+}
+
+// FindNameIndex finds the __name__ label's index
+// if it is different from the most common expected case
+// then it caches the value.
+// The labels slice is expected to be sorted.
+func (tg *trendAsGauges) FindNameIndex() {
```

Comment:
I'd recommend giving another name. The `FindNameIndex` confuses the main purpose of the method as I see is to cache if possible, so my alternatives are something like: `BuildNameIndexCache`, `TryCacheNameIndex`, or something which explicitly tells the purpose of the method.